### ISSUE TITLE
[teoz] [bug] [sequence] Activation is offset when with simultaneous deactivation

### ISSUE THREAD

**Issue by valtusovs** [plantuml/plantuml#1723]
**Describe the bug**
When using teoz engine on the latest plantuml-1.2024.3.jar. When making a sequence diagram. When deactivating actor A and activating actor B, the B's activation box does not start at the arrow.

**To Reproduce**

Copy this file:
```puml
@startuml
!pragma teoz true

dummy -> A ++: first step

A -> A: test
A -> B --++: second step

B -> B: test
B -> C--: third step

@enduml
```

Save the resulting diagram.

Comment out "!pragma teoz true". Save the resulting diagram.

**Expected behavior**
The activation box of actor B starts at the "second step" arrow, like in the puma render.

**Screenshots**
*teoz:*
![image](https://github.com/plantuml/plantuml/assets/50198060/546f2cd4-b86f-4401-8ffc-ee53cb75c7fe)
*puma:*
![image](https://github.com/plantuml/plantuml/assets/50198060/75017741-b676-4b99-b5a7-db9c6e9de17c)

**Comment by jimnelson372** [comment]
I just wanted to let you know that I am looking at these teoz related timeline issues.  I've already found the solution for the --++ condition, but I'm looking at some related problems when & is used.   So, my solution results in the teoz version looking very much like the puma version.  Here's with my changes (not yet submitted):

![testteozlifeline_007](https://github.com/plantuml/plantuml/assets/66284321/d77edc0c-2a3b-4fcc-979b-afdd4db64696)

**Comment by The-Lum** [comment]
Hi @jimnelson372 

Could you have also a look on:
- https://github.com/plantuml/backlog/issues/9

```puml
@startuml
activate a
a -> a : put
deactivate a
@enduml
```
>[![](https://www.planttext.com/api/plantuml/svg/SoWkIImgAStDuKfCBiaiIon9LKZaIbJGjLD8LB1IA2WjuKf9HP9nICrB0Ia60000)](https://www.planttext.com/?text=SoWkIImgAStDuKfCBiaiIon9LKZaIbJGjLD8LB1IA2WjuKf9HP9nICrB0Ia60000) 

_Thanks for your work._
Regards,
Th.

**Comment by jimnelson372** [comment]
Yes @The-Lum,

I had planned to return to the self message lifeline since for forum issue #15191 where the deactivation position changes depending on if it is the last message in the script or not.   

In looking at the backlog#9 discussion, I realize the changes I needed to make to solve the --++ issue has stopped one of the workarounds for extending the self message activation, the use of delay as in:

```
@startuml
!pragma teoz true
activate a
a -> a : put
||0||
deactivate a
@enduml
```
For the --++ fix to work, I had to relax how many instructions away the deactivate occurs, since for --++ and some other conditions the activate could sometimes come before the deactivate and the processing was only looking at the very first of them (which is why --++ wasn't working).  I had to do some additional relaxation in order to fix some of the parallel message issues.   I haven't had a chance to see what it would take to get the ||0|| to force the delay in the deactivation again.

Also, I think we need both kinds of activation level endings on the self message.   I notice that for the slanted messages, such as the following, PlantUML uses the arrow type to determine whether the deactivation should occur at the time the message is sent (if an async arrowhead `->>`), or at the time the message is received (if it is a synchronous arrow head `->` ).

`b ->>(20) a --: async`

Maybe self messages should have the same logic.  Right now, it defaults to the form of a message return, where the source is deactivated and returns to the next lower state, as in:

```
@startuml
!pragma teoz true
a -> a ++: call
a ->> b: an async message happens\nwithin the call
a -->> a --: return
@enduml
```
![testteozselflifeline_001](https://github.com/plantuml/plantuml/assets/66284321/fd13af3c-2855-44fe-8a00-21a3078062dc)

Maybe there should be a new discussion or issue created for this.   I'm not clear what the correct default and solution should be.  Though I do believe I  should restore the effect that ||0|| had prior to my other changes.

Jim N

**Comment by The-Lum** [comment]
Hi all,

_[This is an Issue Review]_ 👀 

This is now fixed on [V1.2024.5](https://github.com/plantuml/plantuml/releases/tag/v1.2024.5).
- Thanks @valtusovs for the report.
- Thanks PlantUML team and @jimnelson372 for your work.

Regards.
